cpp/src/merge/merge.cu

```diff
@@ -0,0 +1,489 @@
+                       auto side = thrust::get<0>(index_pair);
+                       auto index = thrust::get<1>(index_pair);
+                       
+                       ElemenT val = (side == side::LEFT ? p_d_lcol[index] : p_d_rcol[index]);
+                       return val;
+                      }
+                     );
+
+    //cudaDeviceSynchronize();//? nope...these two could proceed concurrently
+
+
+    //CAVEAT: conditional call below is erroneous without
+    //set_null_mask() call (see TODO above):
+    //
+    if (lcol.has_nulls() || rcol.has_nulls())
+      //resolve null mask:
+      //
+      materialize_bitmask(lcol,
+                          rcol,
+                          merged_view,
+                          dv_row_order_.data().get(),
+                          stream_);
+                   
+    return p_merged_col;
+  }
+
+  //specialization for string...?
+  //or should use `cudf::string_view` instead?
+  //
+  template<typename ElemenT>//required: column type
+  std::enable_if_t<not cudf::is_fixed_width<ElemenT>(),
+                   std::unique_ptr<cudf::column>>
+  operator()(cudf::column_view const& lcol, cudf::column_view const& rcol) const
+  {
+    //for now...
+    CUDF_FAIL("Non fixed-width types are not supported");
+
+    // TODO: separate PR for strins support:
+    //
+    // auto column = strings::detail::merge( strings_column_view(lcol),
+    //                                       strings_column_view(rcol),
+    //                                       dv_row_order_.begin(),
+    //                                       dv_row_order_.end(),
+    //                                       mr_,
+    //                                       stream_);
+    
+    // if (lcol.has_nulls() || rcol.has_nulls())
+    //   {
+    //     auto merged_view = column->mutable_view();
+    //     materialize_bitmask(lcol,
+    //                         rcol,
+    //                         merged_view,
+    //                         dv_row_order_.data().get(),
+    //                         stream_);
+    //   }
+    // return column;
+  }
+
+private:
+  VectorI const& dv_row_order_;
+  rmm::mr::device_memory_resource* mr_;
+  cudaStream_t stream_;
+  
+  //see `class element_relational_comparator` in `cpp/include/cudf/table/row_operators.cuh` as a model;
+};
+  
+
+  std::unique_ptr<cudf::experimental::table> merge(cudf::table_view const& left_table,
+                                                   cudf::table_view const& right_table,
+                                                   std::vector<cudf::size_type> const& key_cols,
+                                                   std::vector<cudf::order> const& column_order,
+                                                   std::vector<cudf::null_order> const& null_precedence,
+                                                   rmm::mr::device_memory_resource* mr,
+                                                   cudaStream_t stream = nullptr) {
```

Comment:
```suggestion
                                                   cudaStream_t stream = 0) {
```

Reply:
`cudaStream_t` is a pointer: `typedef CUstream_st * cudaStream_t` 
(https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__TYPES.html#group__CUDART__TYPES_1ge15d9c8b7a240312b533d6122558085a )

To that extent, defaulting it to an int (0), may confuse the compiler (opens the possibility of unintended casts). This is the exact reason why `nullptr` was added to the language, to avoid such confusion. See "Effective Modern C++", S. Meyers, "Item 8: Prefer nullptr to 0 and NULL"

Reply:
cudaStream_t is an opaque type, and the default stream is commonly set to 0 in CUDA. We use `0` everywhere else in libcudf, so I suggest this change to be consistent.

Reply:
Ok, done.